src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/AbstractClassCouplingCheck.java

```diff
@@ -169,28 +217,122 @@ public void leaveToken(DetailAST ast) {
      * @param pkg package definition.
      */
     private void visitPackageDef(DetailAST pkg) {
-        final FullIdent ident = FullIdent.createFullIdent(pkg.getLastChild()
-                .getPreviousSibling());
-        packageName = ident.getText();
+        final FullIdent ident = FullIdent.createFullIdent(pkg.getLastChild().getPreviousSibling());
+        fileContext.setPackageName(ident.getText());
     }
 
     /**
      * Creates new context for a given class.
      * @param classDef class definition node.
      */
     private void visitClassDef(DetailAST classDef) {
-        contextStack.push(context);
-        final String className =
-            classDef.findFirstToken(TokenTypes.IDENT).getText();
-        context = new Context(className,
-                               classDef.getLineNo(),
-                               classDef.getColumnNo());
+        final String className = classDef.findFirstToken(TokenTypes.IDENT).getText();
+        fileContext.createNewClassContext(className, classDef.getLineNo(), classDef.getColumnNo());
     }
 
     /** Restores previous context. */
     private void leaveClassDef() {
-        context.checkCoupling();
-        context = contextStack.pop();
+        fileContext.checkCurrentClassAndRestorePrevious();
+    }
+
+    /**
+     * Encapsulates information about classes coupling inside single file.
+     */
+    private class FileContext {
+        /** A map of (imported class name -> class name with package) pairs. */
+        private final Map<String, String> importedClassPackage = new HashMap<>();
+
+        /** Stack of class contexts. */
+        private final Deque<ClassContext> classesContexts = new ArrayDeque<>();
```

Comment:
Don't understand the question. These classes will be processed as soon as they appear in the ast. I could use a list, but I only need the `push` / `pop` methods.